ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java

```diff
@@ -154,24 +154,22 @@ class MySqlNamedBlobDb implements NamedBlobDb {
    * It will pull out any stale record (limit to be config.queryStaleDataMaxResults [Default 1000] records at most)
    * meeting below conditions:
    * 1. created more than config.staleDataRetentionDays [Default 5] days ago, and
-   * 2. it's NOT a VLR (Valid Latest Record) for any named blob.
+   * 2. its blob_id does NOT show in any VLR (Valid Latest Record) of any named blob.
    * VLR of a named blob is the record whose blob_state=1 (READY), and has the max version for the named blob.
    *
-   * To pull out the rows that meet the 2nd condition above, we are using left join and then pull out the rows where
-   * right side table's keys are all NULL
    */
   // @formatter:off
   private static final String GET_STALE_QUERY = String.format(""
-          + "SELECT t1.%s, t1.%s, t1.%s, t1.%s, t1.%s, t1.%s, %s "
-          + "FROM %s t1 "
-          + "LEFT JOIN "
-          + "(SELECT account_id, container_id, blob_name, max(version) as version "
+          + "SELECT %s, %s, %s, %s, %s, %s, %s "
           + "FROM %s "
-          + "WHERE blob_state=1 "
-          + "      GROUP BY account_id, container_id, blob_name) t2 "
-          + "ON (t1.account_id,t1.container_id,t1.blob_name,t1.version) = (t2.account_id,t2.container_id,t2.blob_name,t2.version) "
-          + "WHERE t1.%s<? and t2.account_id IS NULL and t2.container_id IS NULL and t2.blob_name IS NULL and t2.version IS NULL "
-          + "LIMIT ?",
+          + "WHERE blob_id not in "
+          + "   (SELECT distinct(blob_id) "
+          + "    FROM %s "
+          + "    WHERE version in  (SELECT max(version) as version "
+          + "      FROM %s "
+          + "      WHERE blob_state = 1 and version != 0 "
+          + "      GROUP BY account_id, container_id, blob_name))"
```

Comment:
I would suggest you to add more comments here to explain your ideas.
Something like you select blobId which is not in the subquery, and the subquery is getting distinct blobId where the version == max version of a specific blob.
Also wondering why do we need to order by blobId?
And may I know what's the cost of previous query? We are not an expert for mysql cmd queries, if possible, maybe mysql team can check if this query can be optimized.

Reply:
Ok, just added more comments to make it clear for the ideas behind the SQL query.
I order by blobID so that we have deterministic results within the limit. Eg we have total 50K rows. Each cleanup runner run only pull out the 10K (the default limit number).

The time cost is actually reduced by half (see below). I am asking Bei from mysql team to take a look.

**Cost of previous query:**
```
mysql> SELECT t1.account_id, t1.container_id, t1.blob_name, t1.blob_id, t1.version, t1.deleted_ts, CURRENT_TIMESTAMP(6) FROM named_blobs_v2 t1 LEFT JOIN (SELECT account_id, container_id,
blob_name, max(version) as version  FROM named_blobs_v2  WHERE blob_state=1  GROUP BY account_id, container_id, blob_name) t2   ON (t1.account_id,t1.container_id,t1.blob_name,t1.version) = (t2.account_id,t2.container_id,t2.blob_name,t2.version)   WHERE t1.version<168680160000000000 and t2.account_id IS NULL and t2.container_id IS NULL and t2.blob_name IS NULL and t2.version IS NULL LIMIT 10000;

10000 rows in set (11.59 sec)
```

**Query of latest query with same parameter:**
```
mysql> SELECT account_id, container_id, blob_name, blob_id, version, CURRENT_TIMESTAMP(6) FROM named_blobs_v2 WHERE blob_id not in (SELECT distinct(blob_id)  FROM named_blobs_v2  WHERE version in  (SELECT max(version) as version  FROM named_blobs_v2  WHERE blob_state = 1 and version != 0  GROUP BY account_id, container_id, blob_name) ) AND version<168680160000000000 AND (deleted_ts IS NULL OR deleted_ts>"2023-06-10") ORDER BY blob_id desc LIMIT 10000;

10000 rows in set (6.70 sec)

```